File: app/src/main/java/com/purnendu/tictactoe/MainActivity.java
package com.purnendu.tictactoe;

import androidx.appcompat.app.AppCompatActivity;
import androidx.appcompat.app.AppCompatDelegate;

import android.os.Bundle;
import android.view.View;
import android.widget.RelativeLayout;
import android.widget.TextView;

import com.google.android.material.button.MaterialButton;

class MainActivityJava extends AppCompatActivity implements View.OnClickListener {


    MaterialButton btn1, btn2, btn3, btn4, btn5, btn6, btn7, btn8, btn9, reset;
    TextView playerTurn;
    MaterialButton[][] board;
    int[][] boardStatus;
    int count = 0, turnCount = 0;
    boolean player = true;
    View vm, ht, hb, vr, vl;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        AppCompatDelegate.setDefaultNightMode(AppCompatDelegate.MODE_NIGHT_NO);

        setFindViewById();

        board = new MaterialButton[3][3];
        boardStatus = new int[3][3];

        initializeBoard();

        reset.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                turnCount = 0;
                count = 0;
                player = true;
                playerTurn.setText("Player 1 turn");
                initializeBoard();
                vm.setVisibility(View.INVISIBLE);
                ht.setVisibility(View.INVISIBLE);
                hb.setVisibility(View.INVISIBLE);
                vr.setVisibility(View.INVISIBLE);
                vl.setVisibility(View.INVISIBLE);
                setMiddleVerticalLineToDefaultState();
            }
        });


    }

    private void setFindViewById() {


        playerTurn = findViewById(R.id.playerTurn);
        btn1 = findViewById(R.id.btn1);
        btn2 = findViewById(R.id.btn2);
        btn3 = findViewById(R.id.btn3);
        btn4 = findViewById(R.id.btn4);
        btn5 = findViewById(R.id.btn5);
        btn6 = findViewById(R.id.btn6);
        btn7 = findViewById(R.id.btn7);
        btn8 = findViewById(R.id.btn8);
        btn9 = findViewById(R.id.btn9);
        reset = findViewById(R.id.reset);
        vm = findViewById(R.id.vm);
        ht = findViewById(R.id.ht);
        hb = findViewById(R.id.hb);
        vr = findViewById(R.id.vr);
        vl = findViewById(R.id.vl);

    }

    private MaterialButton getButton(int btnNo) {
        switch (btnNo) {
            case (1):
                return btn1;
            case (2):
                return btn2;
            case (3):
                return btn3;
            case (4):
                return btn4;
            case (5):
                return btn5;
            case (6):
                return btn6;
            case (7):
                return btn7;
            case (8):
                return btn8;
            case (9):
                return btn9;

        }
        return null;
    }

    @Override
    public void onClick(View v) {

        switch (v.getId()) {
            case (R.id.btn1): {
                updateValue(0, 0, player);
                break;
            }
            case (R.id.btn2): {
                updateValue(0, 1, player);
                break;
            }
            case (R.id.btn3): {
                updateValue(0, 2, player);
                break;
            }
            case (R.id.btn4): {
                updateValue(1, 0, player);
                break;
            }
            case (R.id.btn5): {
                updateValue(1, 1, player);
                break;
            }
            case (R.id.btn6): {
                updateValue(1, 2, player);
                break;
            }
            case (R.id.btn7): {
                updateValue(2, 0, player);
                break;
            }
            case (R.id.btn8): {
                updateValue(2, 1, player);
                break;
            }
            case (R.id.btn9): {
                updateValue(2, 2, player);
                break;
            }
        }
        turnCount++;
        player = !player;
        if (player)
            playerTurn.setText("Player X turn");
        else
            playerTurn.setText("Player 0 turn");

        if (turnCount == 9)
            playerTurn.setText("Game Draw");

        checkWinner();

    }


    private void initializeBoard() {

        for (int i = 0; i < 3; i++) {
            for (int j = 0; j < 3; j++) {
                count++;
                MaterialButton btn = getButton(count);
                if (btn != null) {
                    board[i][j] = btn;
                    board[i][j].setOnClickListener(this);
                    boardStatus[i][j] = -1;
                    board[i][j].setEnabled(true);
                    board[i][j].setText("");
                }

            }
        }
    }

    private void updateValue(int row, int column, boolean player) {
        String temp = player ? "X" : "0";
        int tempInt = player ? 1 : 0;
        board[row][column].setText(temp);
        boardStatus[row][column] = tempInt;
        board[row][column].setEnabled(false);
    }

    private void checkWinner() {

        String line = "";

        //Horizontally Check
        for (int i = 0; i < 3; i++) {
            if ((boardStatus[i][0] == boardStatus[i][1]) && (boardStatus[i][0] == boardStatus[i][2])) {
                if (boardStatus[i][0] == 1) {
                    playerTurn.setText("Player 1 win");
                    line = "row" + ":" + i;
                    break;
                } else if (boardStatus[i][0] == 0) {
                    playerTurn.setText("Player 2 win");
                    line = "row" + ":" + i;
                    break;
                }

            }
        }

        //Vertically Check
        for (int i = 0; i < 3; i++) {
            if ((boardStatus[0][i] == boardStatus[1][i]) && (boardStatus[0][i] == boardStatus[2][i])) {
                if (boardStatus[0][i] == 1) {
                    playerTurn.setText("Player 1 win");
                    line = "column" + ":" + i;
                    break;
                } else if (boardStatus[0][i] == 0) {
                    playerTurn.setText("Player 2 win");
                    line = "column" + ":" + i;
                    break;
                }
            }
        }

        //Diagonally Check
        if ((boardStatus[0][0] == boardStatus[1][1]) && (boardStatus[0][0] == boardStatus[2][2])) {
            if (boardStatus[0][0] == 1) {
                line = "corner:l";
                playerTurn.setText("Player 1 win");
            } else if (boardStatus[0][0] == 0) {
                line = "corner:l";
                playerTurn.setText("Player 2 win");
            }


        }

        if ((boardStatus[0][2] == boardStatus[1][1]) && (boardStatus[0][2] == boardStatus[2][0])) {
            if (boardStatus[0][2] == 1) {
                playerTurn.setText("Player 1 win");
                line = "corner:r";
            } else if (boardStatus[0][2] == 0) {
                playerTurn.setText("Player 2 win");
                line = "corner:r";
            }
        }
        lineDraw(line);
        disableButton();
    }

    private void lineDraw(String line) {
        if (!line.isEmpty()) {
            switch (line) {
                case "row:0":
                    ht.setVisibility(View.VISIBLE);
                    break;
                case "row:1": {
                    int buttonWidth = btn3.getWidth();
                    int width = (buttonWidth * 3);
                    vm.setRotation(90);
                    vm.getLayoutParams().height = width;
                    vm.setVisibility(View.VISIBLE);
                    break;
                }
                case "row:2":
                    hb.setVisibility(View.VISIBLE);
                    break;
                case "corner:l":
                    vm.setRotation(-33);
                    vm.setVisibility(View.VISIBLE);
                    break;
                case "corner:r":
                    vm.setRotation(33);
                    vm.setVisibility(View.VISIBLE);
                    break;
                case "column:0": {
                    int buttonWidth = btn1.getWidth();
                    int width = (buttonWidth / 2) + 20;
                    RelativeLayout.LayoutParams params = (RelativeLayout.LayoutParams) vr.getLayoutParams();
                    params.setMargins(width, 200, 0, 200); //left, top, right, bottom
                    vr.setLayoutParams(params);
                    vr.setVisibility(View.VISIBLE);
                    break;
                }
                case "column:1": {
                    RelativeLayout.LayoutParams params = (RelativeLayout.LayoutParams) vm.getLayoutParams();
                    params.setMargins(0, 200, 0, 200); //left, top, right, bottom
                    vm.setLayoutParams(params);
                    vm.setVisibility(View.VISIBLE);
                    break;
                }
                case "column:2": {
                    int buttonWidth = btn3.getWidth();
                    int width = (buttonWidth / 2) + 20;
                    RelativeLayout.LayoutParams params = (RelativeLayout.LayoutParams) vl.getLayoutParams();
                    params.setMargins(0, 200, width, 200); //left, top, right, bottom
                    vl.setLayoutParams(params);
                    vl.setVisibility(View.VISIBLE);
                    break;
                }
            }
        }
    }

    private void disableButton() {

        if (playerTurn.getText().toString().contains("win")) {
            for (int i = 0; i < 3; i++) {
                for (int j = 0; j < 3; j++) {
                    board[i][j].setEnabled(false);
                }
            }
        }
    }

    private void setMiddleVerticalLineToDefaultState() {
        vm.setRotation(0);
        int buttonWidth = btn2.getHeight();
        int height = (buttonWidth * 3);
        vm.getLayoutParams().height = height;
    }

}

File: app/src/main/java/com/purnendu/tictactoe/MainActivity.kt
package com.purnendu.tictactoe

import android.os.Bundle
import android.view.View
import android.widget.RelativeLayout
import androidx.appcompat.app.AppCompatActivity
import androidx.appcompat.app.AppCompatDelegate
import androidx.databinding.DataBindingUtil
import com.google.android.material.button.MaterialButton
import com.purnendu.tictactoe.databinding.ActivityMainBinding

class MainActivity : AppCompatActivity(), View.OnClickListener {


    private lateinit var binding: ActivityMainBinding
    private lateinit var board: Array<Array<MaterialButton>>
    private lateinit var boardStatus: Array<Array<Int>>


    private var count: Int = 0
    private var turnCount: Int = 0
    private var player: Boolean = true


    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        binding = DataBindingUtil.setContentView(this, R.layout.activity_main)

        //Set default night mode -> no
        AppCompatDelegate.setDefaultNightMode(AppCompatDelegate.MODE_NIGHT_NO)


        //Initialization of Arrays
        board = arrayOf(
            arrayOf(binding.btn1, binding.btn2, binding.btn3),
            arrayOf(binding.btn4, binding.btn5, binding.btn6),
            arrayOf(binding.btn7, binding.btn8, binding.btn9)
        )

        boardStatus = arrayOf(
            arrayOf(0, 0, 0),
            arrayOf(0, 0, 0),
            arrayOf(0, 0, 0)
        )

        initializeBoard()


        binding.reset.setOnClickListener {

            turnCount = 0
            count = 0
            player = true
            binding.playerTurn.text = "Player X turn"
            initializeBoard()
            binding.vm.visibility = View.INVISIBLE
            binding.ht.visibility = View.INVISIBLE
            binding.hb.visibility = View.INVISIBLE
            binding.vr.visibility = View.INVISIBLE
            binding.vl.visibility = View.INVISIBLE
            setMiddleVerticalLineToDefaultState()
        }

    }

    private fun initializeBoard() {

        for (i in 0..2) {
            for (j in 0..2) {
                count++
                board[i][j].setOnClickListener(this)
                boardStatus[i][j] = -1
                board[i][j].isEnabled = true
                board[i][j].text = ""

            }
        }
    }

    private fun updateValue(row: Int, column: Int, player: Boolean) {
        val temp: String = if (player) "X" else "0"
        val tempInt: Int = if (player) 1 else 0
        board[row][column].text = temp
        boardStatus[row][column] = tempInt
        board[row][column].isEnabled = false
    }

    private fun checkWinner() {
        var line = ""

        //Horizontally Check
        for (i in 0..2) {
            if ((boardStatus[i][0] == boardStatus[i][1]) && (boardStatus[i][0] == boardStatus[i][2])) {
                if (boardStatus[i][0] == 1) {
                    binding.playerTurn.text = "Player X win"
                    line = "row:$i"
                    break
                } else if (boardStatus[i][0] == 0) {
                    binding.playerTurn.text = "Player 0 win"
                    line = "row:$i"
                    break
                }

            }
        }

        //Vertically Check
        for (i in 0..2) {
            if ((boardStatus[0][i] == boardStatus[1][i]) && (boardStatus[0][i] == boardStatus[2][i])) {
                if (boardStatus[0][i] == 1) {
                    binding.playerTurn.text = "Player X win"
                    line = "column:$i"
                    break
                } else if (boardStatus[0][i] == 0) {
                    binding.playerTurn.text = "Player 0 win"
                    line = "column:$i"
                    break
                }
            }
        }

        //Diagonally Check
        if ((boardStatus[0][0] == boardStatus[1][1]) && (boardStatus[0][0] == boardStatus[2][2])) {
            if (boardStatus[0][0] == 1) {
                line = "corner:l"
                binding.playerTurn.text = "Player X win"
            } else if (boardStatus[0][0] == 0) {
                line = "corner:l"
                binding.playerTurn.text = "Player 0 win"
            }


        }

        if ((boardStatus[0][2] == boardStatus[1][1]) && (boardStatus[0][2] == boardStatus[2][0])) {
            if (boardStatus[0][2] == 1) {
                binding.playerTurn.text = "Player X win"
                line = "corner:r"
            } else if (boardStatus[0][2] == 0) {
                binding.playerTurn.text = "Player 0 win"
                line = "corner:r"
            }
        }
        lineDraw(line)
        disableButton()

    }

    private fun lineDraw(line: String) {
        if (line.isNotEmpty()) {

            when (line) {
                "row:0" -> binding.ht.visibility = View.VISIBLE
                "row:1" -> {
                    val buttonWidth: Int = binding.btn3.width
                    val width: Int = (buttonWidth * 3)
                    binding.vm.apply {
                        rotation = 90F
                        layoutParams.height = width
                        visibility = View.VISIBLE
                    }
                }

                "row:2" -> binding.hb.visibility = View.VISIBLE
                "corner:l" -> {
                    binding.vm.apply {
                        rotation = -33F
                        visibility = View.VISIBLE
                    }
                }
                "corner:r" -> {
                    binding.vm.apply {
                        rotation = 33F
                        visibility = View.VISIBLE
                    }
                }
                "column:0" -> {
                    val buttonWidth: Int = binding.btn1.width
                    val width: Int = (buttonWidth / 2) + 20
                    val params: RelativeLayout.LayoutParams =
                        binding.vr.layoutParams as RelativeLayout.LayoutParams
                    params.setMargins(width, 200, 0, 200)//left, top, right, bottom
                    binding.vr.apply {
                        layoutParams = params
                        visibility = View.VISIBLE
                    }
                }
                "column:1" -> {
                    val params: RelativeLayout.LayoutParams =
                        binding.vm.layoutParams as RelativeLayout.LayoutParams
                    params.setMargins(0, 200, 0, 200) //left, top, right, bottom
                    binding.vm.apply {
                        layoutParams = params
                        visibility = View.VISIBLE
                    }
                }
                "column:2" -> {
                    val buttonWidth: Int = binding.btn3.width
                    val width: Int = (buttonWidth / 2) + 20
                    val params: RelativeLayout.LayoutParams =
                        binding.vl.layoutParams as RelativeLayout.LayoutParams
                    params.setMargins(0, 200, width, 200)//left, top, right, bottom
                    binding.vl.apply {
                        layoutParams = params
                        visibility = View.VISIBLE
                    }
                }
            }

        }

    }

    private fun disableButton() {

        if (binding.playerTurn.text.toString().contains("win")) {
            for (i in 0..2) {
                for (j in 0..2) {
                    board[i][j].isEnabled = false
                }
            }
        }

    }

    private fun setMiddleVerticalLineToDefaultState() {

        binding.vm.rotation = 0F
        val buttonWidth: Int = binding.btn2.height
        val height: Int = (buttonWidth * 3)
        binding.vm.layoutParams.height = height
    }


    override fun onClick(view: View) {

        when (view.id) {

            (R.id.btn1) -> updateValue(0, 0, player)
            (R.id.btn2) -> updateValue(0, 1, player)
            (R.id.btn3) -> updateValue(0, 2, player)
            (R.id.btn4) -> updateValue(1, 0, player)
            (R.id.btn5) -> updateValue(1, 1, player)
            (R.id.btn6) -> updateValue(1, 2, player)
            (R.id.btn7) -> updateValue(2, 0, player)
            (R.id.btn8) -> updateValue(2, 1, player)
            (R.id.btn9) -> updateValue(2, 2, player)
        }

        turnCount++
        player = !player

        if (player) binding.playerTurn.text = "Player X turn" else binding.playerTurn.text =
            "Player 0 turn"

        if (turnCount == 9) binding.playerTurn.text = "Game Draw"

        checkWinner()
    }
}